### ISSUE TITLE
Timing test for undocumented AXA (zp),Y

### ISSUE THREAD

**Issue by tomcw** [AppleWin/AppleWin#282]
There's a comment in AppleWin source saying:

````
#define AXA	 /*bSlowerOnPagecross = 0;*//*FIXME: $93 case is still unclear*/
````

Test code:

300: A0 FF 84 FE A0 20 84 FF A0 00 AD 04 C4 93 FE AE 04 C4 00

````
.ORG $300
 LDY #$FF
 STY $FE
 LDY #$20
 STY $FF
 LDY #0 ; For page-cross: 309:1
 LDA $C404
 AXA ($FE),Y
 LDX $C404
 BRK
````

NB. Below, delta = (A-X) - 4 cycles for the LDX $C404

AppleWin v1.25.0.3:
* Y=0 (no page cross): delta = 6
* Y=1 (page cross): delta = 6

Original Apple //e:
* Y=0 (no page cross): delta = 6
* Y=1 (page cross): delta = 6

**EDIT:** After seeing than the page-crossing case of AXA abs,Y doesn't update the address abs+Y I checked this opcode $93, and get the same result: page-crossing doesn't update the target address (but non-page crossing does).

**Comment by tomcw** [comment]
Also for completeness I tested: AXA abs,Y

300: A0 00 A2 00 AD 04 C4 9F FF 20 AE 04 C4 00

````
.ORG $300
 LDY #0 ; For page-cross: 301:1
 LDX #0 ; Zero here to force AXA result to be zero
 LDA $C404
 AXA $20FF,Y
 LDX $C404
 BRK
````

NB. Below, delta = (A-X) - 4 cycles for the LDA $C404

AppleWin v1.25.0.3:
* Y=0 (no page cross): delta = 5
* Y=1 (page cross): delta = 5

Original Apple //e:
* Y=0 (no page cross): delta = 5 (OK, stores zero to $20FF)
* Y=1 (page cross): delta = 5 (**$2100 remains unchanged!**)

NB. With the page-crossing case I don't know if (or where) the result gets written.
I filled memory $2000..$2FFF with FF's (2001<2000.3000M), but for the Y=1 case (AXA result of 0) the memory was still FF's. I repeated by filling the same memory with 11's in case AXA was writing FF, but still nothing was written.

**Comment by sicklittlemonkey** [comment]
Interesting. That Mockingboard timer is handy!

I don't have an NMOS 6502 machine working to have a look.

You might be able to get some more insight using the gate-level Visual 6502
simulator.

Here's a URL with your test program put at $300 and trace run for 28 steps.
The $9F is executed from cycle 8 onwards.
http://visual6502.org/JSSim/expert.html?graphics=f&r=0300&a=0300&d=A000A200AD04C49FFF20AE04C400&steps=28&loglevel=4&logmore=adl,adh,idl,idb,alua,alub,alu,dasb,sb,dpc2_XSB,dpc9_DBADD,dpc11_SBADD,dpc23_SBAC,dpc24_ACSB,dpc25_SBDB,dpc43_DL/DB,DPControl

Here are the docs for the URL API:
http://visual6502.org/wiki/index.php?title=JssimUserHelp

Note that some opcodes really do appear to be unstable:
http://visual6502.org/wiki/index.php?title=6502_Opcode_8B_(XAA,_ANE)

There's also an interesting auto-generated summary which might or might not
be accurate:
http://visual6502.org/wiki/index.php?title=6502_all_256_Opcodes

Cheers,
Nick.

**Comment by tomcw** [comment]
Wow! That Visual6502 simulator looks very neat!

I had considered sticking a scope on the 6502's R/W pin to detect the presence (or absence) of a write to memory.

Anyway thanks for the links.

**Comment by tomcw** [comment]
http://www.ffd2.com/fridge/docs/6502-NMOS.extra.opcodes

The last warning (below) may explain why I'm not seeing the write, and in fact it is occurring to zero-page instead.

````
The following notes apply to the above four opcodes: TAS, SAY, XAS, AXA.

None of these opcodes affect the accumulator, the X register, the Y 
register, or the processor status register!
     The author has no explanation for the complexity of these 
instructions.  It is hard to comprehend how the microprocessor could handle 
the convoluted sequence of events which appears to occur while executing 
one of these opcodes.  A partial explanation for what is going on is that 
these instructions appear to be corruptions of other instructions.  For 
example, the opcode SAY would have been one of the addressing modes of the 
standard instruction STY (absolute indexed X) were it not for the fact that 
the normal operation of this instruction is impaired in this particular 
instance.

One irregularity uncovered is that sometimes the actual value is stored in 
memory, and the AND with <ab+1> part drops off (ex. SAY becomes true STY).  
This happens very infrequently.  The behaviour appears to be connected with 
the video display.  For example, it never seems to occur if either the 
screen is blanked or C128 2MHz mode is enabled.

--- Imported example ---
Here is a demo program to illustrate the above effect.  SYS 8200 to try it.  
There is no exit, so you'll have to hit Stop-Restore to quit.  And you may 
want to clear the screen before running it.  For contrast, there is a 
second routine which runs during idle state display.  Use SYS 8211 for it.  
After trying the second routine, check it out again using POKE 53269,255 to 
enable sprites.

WARNING: If the target address crosses a page boundary because of indexing, 
the instruction may not store at the intended address.  It may end up 
storing in zero page, or another address altogether (page=value stored). 
Apparently certain internal 65XX registers are being overridden.  The whole 
scheme behind this erratic behaviour is very complex and strange.  
````

**Comment by tomcw** [comment]
# AXA abs,Y

**Non-page-crossing case:**
- Y = #$41
- AXA $4000,Y

A | X | base+1 | result | address
--- | --- | --- | --- | ---
EA | 20 | 41 | 00 | $4041
EA | FF | 41 | 40 | $4041
FF | FF | 41 | 41 | $4041

**Page-crossing case:**
- Y = #$41
- AXA $40FF,Y

A | X | base+1 | result | address
--- | --- | --- | --- | ---
EA | 20 | 41 | 00 | $0040
EA | FF | 41 | 40 | $4040
FF | FF | 41 | 41 | $4140

So for the page-crossing case, the result is used as the high-byte of the store address.

This result matches the WARNING in the preceding comment, ie:
> It may end up storing in zero page, or another address altogether (page=value stored).

----

For reference my test code:
(copies & restores ZP to $2000)

300:A5 00 8D 00 20 EE 01 03 EE 03 03 D0 F3 A9 EA 85 00 EE 10 03 D0 F9 A2 FF A0 41 9F FF 40 A5 00 8D 00 21 EE 1E 03 EE 20 03 D0 F3 AD 00 20 85 00 EE 2B 03 EE 2E 03 D0 F3 60

````
300:	; memcpy($2000, $00, $100)
 lda $00
 sta $2000
 inc $301
 inc $303
 bne $300

 lda #$EA
30F:	; memset($00, #$EA, $100)
 sta $0
 inc $310
 bne $30F

 ldx #$FF
 ldy #$41
 axa $40FF,y

31B:	; memcpy($2100, $00, $100)
 lda $00
 sta $2100
 inc $31C
 inc $31E
 bne $31B

328:	; memcpy($00, $2000, $100)
 lda $2000
 sta $00
 inc $329
 inc $32C
 bne $300

 rts
````

**Comment by tomcw** [comment]
NB. I checked AXA (zp),Y on real h/w and this behaves the same as AXA abs,Y when page-crossing.

**Comment by tomcw** [comment]
Fixed in 614dfc1257cc7fc32fb344be27944f145cdd8935. Closing.
